### PR TITLE
Training algorithm and dockerfile implementation

### PR DESCRIPTION
**Pull request by kerrlabajo**
This PR will be responsible in the following implementation:
1. Implementation of the complete training algorithm using the `yolov5`.
2. Dataset's `.rar` or `.zip` extraction and retrieval.
4. Executing the algorithm in the form of an `Ubuntu image` that can accept parameters.
5. Saving the model in **AWS S3**.
6. Locally running the `Ubuntu image` with the algorithm and its parameters in a `Docker container` for testing.
7. Pushing the `algorithm image` to the **AWS ECR** to be used for PR #1.

> Upon finalizing this implementation, the team will have to undergo several testing with the `Training Job creation in AWS Console Management` (Web Interface) before proceeding with the **.NET development** in PR #1.

References:
https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html
https://github.com/aws/sagemaker-training-toolkit
https://github.com/aws/amazon-sagemaker-examples/tree/main/advanced_functionality/custom-training-containers

### CODE REVIEW

**Comment by kerrlabajo** [comment]
${{\color{Green}\Huge{\textsf{  Implementation\ Updates:\ \}}}}\$
1. Dataset download and file unzipping successful.
2. Manual `data.yaml` modification.
3. Running the training but still with predefined parameters.
4. Saving the weights in my AWS Account's **S3 bucket** specified to **SageMaker**.
4. `Dockerfile` building and running successfully in local.

${{\color{Red}\Huge{\textsf{  Remaining\ Tasks:\ \}}}}\$
1. Parameters must accept parseable arguments from `Dockerfile` entrypoint/cmd ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
2. Executing the `build_and_push.sh script` ${{\color{Green}{\textsf{  [EXECUTABLE\ ANYTIME]\ \}}}}\$.
3. Creating a _training job_ in **AWS Management Console** (or Web Interface) ${{\color{Red}{\textsf{  [ONLY\ AFTER\ BUILD\ AND\ PUSH]\ \}}}}\$.
4. Finding out if _training job creation_ can accept arguments/parameters ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
5. Exporting `best.pt` to `.onnx` format using `export.py` in yolov5 before saving it in **S3 bucket** ${{\color{Red}{\textsf{  [TO\ BE\ IMPLEMENTED]\ \}}}}\$.

**Comment by kerrlabajo** [comment]
${{\color{Green}\Huge{\textsf{  Implementation\ Updates:\ \}}}}\$
1. Predefine expected paths for SageMaker.
2. Dataset download and file unzipping successful.
3. Manual `data.yaml` modification.
4. Running the training but still with predefined parameters.
5. Saving the **ONNX** weights in my AWS Account's **S3 bucket** specified to **SageMaker**.
4. `Dockerfile` building and running successfully in local.
6. Relocate functionalities in their methods.
7. Add action timer per methods that was executed for logging purposes.

${{\color{Red}\Huge{\textsf{  Remaining\ Tasks:\ \}}}}\$
1. Parameters must accept parseable arguments from `Dockerfile` entrypoint/cmd ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
2. Executing the `build_and_push.sh script` ${{\color{Green}{\textsf{  [EXECUTABLE\ ANYTIME]\ \}}}}\$.
3. Creating a _training job_ in **AWS Management Console** (or Web Interface) ${{\color{Red}{\textsf{  [ONLY\ AFTER\ BUILD\ AND\ PUSH]\ \}}}}\$.
4. Finding out if _training job creation_ can accept arguments/parameters ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
5. ~~Exporting `best.pt` to `.onnx` format using `export.py` in yolov5 before saving it in **S3 bucket**~~ ${{\color{Green}{\textsf{  [COMPLETED]\ \}}}}\$.

**Comment by kerrlabajo** [comment]
${{\color{Green}\Huge{\textsf{  Implementation\ Updates:\ \}}}}\$
1. Predefine expected paths for SageMaker.
2. Dataset download and file unzipping successful.
3. Manual `data.yaml` modification.
4. Running the training but still with predefined parameters.
5. Saving the **ONNX** weights in my AWS Account's **S3 bucket** specified to **SageMaker**.
4. `Dockerfile` building and running successfully in local.
6. Relocate functionalities in their methods.
7. Add action timer per methods that was executed for logging purposes.
8. Fix python dependency/ies.
9. Configure `build_and_push.sh` script to push image in **AWS ECR**.

${{\color{Red}\Huge{\textsf{  Remaining\ Tasks:\ \}}}}\$
1. Parameters must accept parseable arguments from `Dockerfile` entrypoint/cmd ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
2. ~~Executing the `build_and_push.sh script`~~ ${{\color{Green}{\textsf{  [SUCCESSFUL]\ \}}}}\$.
3. ~~Creating a _training job_ in **AWS Management Console** (or Web Interface)~~ ${{\color{Green}{\textsf{  [SUCCESSFUL]\ \}}}}\$
image
image

4. Finding out if _training job creation_ can accept arguments/parameters ${{\color{Orange}{\textsf{  [IN\ PROGRESS]\ \}}}}\$.
5. ~~Exporting `best.pt` to `.onnx` format using `export.py` in yolov5 before saving it in **S3 bucket**~~ ${{\color{Green}{\textsf{  [COMPLETED]\ \}}}}\$.

**Comment by kerrlabajo** [comment]
In my observation, this [link](<https://github.com/aws/amazon-sagemaker-examples/tree/main/advanced_functionality/pytorch_yolov5_training_and_hpo>) and their explanation at the 15th cell of the jupyter notebook, using `yolov5s/n/n6 or etc.` as a model architecture to train a custom dataset is called ***Transfer Learning***.

The aforementioned link above and this following [link](<https://github.com/aws/amazon-sagemaker-examples/blob/main/advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb>) as well as this [link ](https://github.com/aws-samples/amazon-sagemaker-train-and-deploy-yolov5)expressed similarity in terms of training approaches and should be the **__core references__** to our study.

These references only differ from our study in terms of:
1. The environment that these notebooks must be run on SageMaker notebook instances.
2. Inside the notebook, a particular algorithm shows the usage of hyperparameters.json/dict for custom parameters to be *automatically* injected to the `yolov5/train.py` script using the `sagemaker.pytorch`.
3. The notebook also handles the training job creation.
image

Both of the references had Dockerfiles but did not have particular ENTRYPOINTS or CMDS therefore it was being carried out instead on the first image. IF that was the case, therefore, there was no need to create another train.py script that handles dataset download and extraction, yaml modification, and executing a subprocess that depends on the yolov5/train.py BUT instead, depend on the .NET Training Job creation implementation on how to handle such feature from the intended script and directly use yolov5/train.py whilst using a generated hyperparameters.json to be injected.

In conclusion, further editing this `Dockerfile` and `train.py` will be discontinued and instead, transfer these features to the .NET SageMaker Training Job Creation Development.

**Comment by kerrlabajo** [comment]
In addition to the previous comment, the following structure shown below that was utilized in this branch's `train.py` was not entirely used even though there were output results saved properly in their output path and did not reflect immediately in S3.
image

Therefore, this structure must have only been usable if it was a `sagemaker.pytorch.fit()` execution or the Training Job Creation in .NET's AWSSDK.SageMaker Implementation.

**Comment by kerrlabajo** [comment]
Before closing this PR, the branch `dev/dockerize-algorithm` should not be deleted but locked instead for progress safekeeping. The next dev should be in a different branch that will still deal with a different `Dockerfile` but without the custom `train.py` script.

**Comment by kerrlabajo** [comment]
Before closing this PR, the branch `dev/dockerize-algorithm` should not be deleted but locked instead for progress safekeeping. The next dev should be in a different branch that will still deal with a different `Dockerfile` but without the custom `train.py` script.
